src/algebra/star/chsh.lean

```diff
@@ -0,0 +1,265 @@
+which we hide in a namespace as they are unlikely to be useful elsewhere.
+-/
+local notation `√2` := (real.sqrt 2 : ℝ)
+
+namespace tsirelson_inequality
+
+/-!
+We next need some lemmas about numerals in modules and algebras.
+The awkward appearance of both `•ℤ` and `•` seems unavoidable because later calculations by `abel`
+will introduce `•ℤ`.
+If anyone sees how to obtain these from general statements, please improve this!
+-/
+
+lemma two_gsmul_half_smul {α : Type*} [add_comm_group α] [module ℝ α] {X : α} :
+  2 •ℤ (2⁻¹ : ℝ) • X = X :=
+by { rw [gsmul_eq_smul_cast ℝ, ←mul_smul]; norm_num, }
+
+lemma neg_two_gsmul_half_smul {α : Type*} [add_comm_group α] [module ℝ α] {X : α} :
+  (-2) •ℤ (2⁻¹ : ℝ) • X = - X :=
+by { rw [gsmul_eq_smul_cast ℝ, ←mul_smul]; norm_num, }
+
+lemma smul_two {α : Type*} [ring α] [algebra ℝ α] {x : ℝ} :
+  x • (2 : α) = (2 * x) • 1 :=
+by { rw [mul_comm 2 x, mul_smul], simp, }
+
+lemma smul_four {α : Type*} [ring α] [algebra ℝ α] {x : ℝ} :
+  x • (4 : α) = (4 * x) • 1 :=
+by { rw [mul_comm 4 x, mul_smul], simp, }
+
+/-!
+Before proving Tsirelson's bound,
+we prepare some easy lemmas about √2.
+-/
+
+-- This calculation, which we need for Tsirelson's bound,
+-- defeated me. Thanks for the rescue from Shing Tak Lam!
+lemma tsirelson_inequality_aux : √2 * √2 ^ 3 = √2 * (2 * √2⁻¹ + 4 * (√2⁻¹ * 2⁻¹)) :=
+begin
+  ring,
+  rw [mul_assoc, inv_mul_cancel, real.sqrt_eq_rpow, ←real.rpow_nat_cast, ←real.rpow_mul],
+  { norm_num,
+    rw show (2 : ℝ) ^ (2 : ℝ) = (2 : ℝ) ^ (2 : ℕ), by { rw ←real.rpow_nat_cast, norm_num },
+    norm_num },
+  { norm_num, },
+  { norm_num, },
+end
+
+lemma sqrt_two_inv_sq : √2⁻¹ * √2⁻¹ = (2⁻¹ : ℝ) :=
+by { rw [←mul_inv'], norm_num, }
+
+end tsirelson_inequality
+open tsirelson_inequality
+
+/--
+In a noncommutative ordered *-algebra over ℝ,
+Tsirelson's bound for a CHSH tuple (A₀, A₁, B₀, B₁) is
+`A₀ * B₀ + A₀ * B₁ + A₁ * B₀ - A₁ * B₁ ≤ 2^(3/2) • 1`.
+
+We prove this by providing an explicit sum-of-squares decomposition
+of the difference.
+
+(We could work over `ℤ[2^(1/2), 2^(-1/2)]` if we really wanted to!)
+-/
+lemma tsirelson_inequality
+  [ordered_ring R] [star_ordered_ring R]
+  [algebra ℝ R] [ordered_semimodule ℝ R] [star_algebra ℝ R]
+  (A₀ A₁ B₀ B₁ : R) (T : is_CHSH_tuple A₀ A₁ B₀ B₁) :
+  A₀ * B₀ + A₀ * B₁ + A₁ * B₀ - A₁ * B₁ ≤ √2^3 • 1 :=
+begin
+  let P := √2⁻¹ • (A₁ + A₀) - B₀,
+  let Q := √2⁻¹ • (A₁ - A₀) + B₁,
+  have w : √2^3 • 1 - A₀ * B₀ - A₀ * B₁ - A₁ * B₀ + A₁ * B₁ = √2⁻¹ • (P^2 + Q^2),
+  { dsimp [P, Q],
+    simp only [pow_two, sub_mul, mul_sub, add_mul, mul_add, smul_add, smul_sub],
+    simp only [algebra.mul_smul_comm, algebra.smul_mul_assoc, ←mul_smul, sqrt_two_inv_sq],
+    simp only [←pow_two, T.A₀_inv, T.A₁_inv, T.B₀_inv, T.B₁_inv],
+    simp only [←T.A₀B₀_commutes, ←T.A₀B₁_commutes, ←T.A₁B₀_commutes, ←T.A₁B₁_commutes],
+    abel,
+    simp only [two_gsmul_half_smul, neg_two_gsmul_half_smul],
+    abel,
```

Comment:
Would it be possible to put some of this in `calc` blocks, or is there too much stuff happening "on both sides"?

Reply:
I was disinclined to make this readable since a grobner basis tactic should kill it easily, so humans really shouldn't be reading it. :-)

Reply:
That's fair, but I suspect we'll have to port this to Lean 4 before we get a Gröbner basis tactic and having an easier to read proof could potentially help then.

Reply:
It's all "one-sided", but the actual expressions are pretty gross, so I don't think `calc` is going to be very illuminating. I guess we could do a calc block with lots of `_`s.

In the meantime, I've added a comment on each line explaining what's going on structurally. If this looks okay I'd prefer to go with that.